module/zfs/zvol.c

```diff
@@ -367,12 +393,16 @@ zvol_set_volsize(const char *name, uint64_t volsize)
 
 	mutex_enter(&zvol_state_lock);
 	zv = zvol_find_by_name(name);
+	if (zv != NULL)
+		mutex_enter(&zv->zv_state_lock);
+	mutex_exit(&zvol_state_lock);
 
 	if (zv == NULL || zv->zv_objset == NULL) {
 		if ((error = dmu_objset_own(name, DMU_OST_ZVOL, B_FALSE,
 		    FTAG, &os)) != 0) {
-			mutex_exit(&zvol_state_lock);
-			return (SET_ERROR(error));
+			if (zv != NULL)
+				mutex_exit(&zv->zv_state_lock);
+			return (SET_ERROR(ENXIO));
```

Comment:
Did you find an issue here returning `error` instead of `ENXIO` unconditionally?